src/builtins/commandline.cpp

```diff
@@ -306,17 +249,8 @@ maybe_t<int> builtin_commandline(parser_t &parser, io_streams_t &streams, const
                     if (ld.is_repaint) continue;
                 }
 
-                // HACK: Execute these right here and now so they can affect any insertions/changes
-                // made via bindings. The correct solution is to change all `commandline`
-                // insert/replace operations into readline functions with associated data, so that
-                // all queued `commandline` operations - including buffer modifications - are
-                // executed in order
-                if (mc == rl::begin_undo_group || mc == rl::end_undo_group) {
-                    reader_handle_command(*mc);
-                } else {
-                    // Inserts the readline function at the back of the queue.
-                    reader_queue_ch(*mc);
-                }
+                // Inserts the readline function at the back of the queue.
+                reader_queue_ch(readline_cmd_t(*mc));
```

Comment:
BTW we currently don't use undo groups.

Our plan is to make it so that every key press either
0. odes not modify undo history
1. adds to an existing undo entry (when adding to the current word)
2. or creates exactly one new undo entry (which needs the undo group if there are multiple changes)

This way undo groups need not be user visible. That's #7885